Implement and test file seek, extract common "read_exactly" for reuse

BenWiederhake opened the pull request:
This PR:
- Extracts the private method `osmium::io::detail::PbfParser::read_exactly` as a generic, easily reusable function
- Implements `osmium::util::file_seek` for random access in files (currently unused, see below)
- Tests the seek feature (which also uses `read_exactly`)

This works toward #367, which is where `file_seek` will be very useful.

If you think `file_seek` should instead live in `experimental`, I'd be happy to move the function there.

BenWiederhake commented:
Changes since last push:
- Added `#include <cassert>` to read_write.hpp, since I use assert(). Found through iwyu, even though its output is incredibly noisy.

joto commented on include/osmium/util/file.hpp:
```diff
@@ -222,6 +222,22 @@ namespace osmium {
+#ifdef _MSC_VER
+            osmium::detail::disable_invalid_parameter_handler diph;
+            // https://msdn.microsoft.com/en-us/library/1yee101t.aspx
+            _lseeki64(fd, offset, SEEK_SET);
```
We'll need a cast here (and below for the `lseek`) because `off_t` is a signed type and `size_t` is unsigned, to avoid warnings.

joto commented on test/t/io/test_file_seek.cpp:
```diff
@@ -0,0 +1,62 @@
+        { 0x21, {0x56, 0xc6, 0x18, 0xc3, 0xea, 0x6d, 0x4f, 0xe0} }, /* "long" backward jump */
+    };
+    for (const auto& expect : expectations) {
+        printf("Trying to seek to %lu\n", expect.offset);
```
Please remove this line, tests should usually not produce output.

joto reviewed (changes requested):
Looks good, except the two small things I wrote down.

I am getting some weird effects here that github actions seems not to be testing your newsest commit. Hopefully this sorts itself out when you add those changes.

BenWiederhake commented:
Changes since last push:
- Added the cast to __int64, that seems to be the type that's also used in other MSVC-specific implementations, like resize_file.
- Removed the print from the test
